src/collectors/mdstat/mdstat.py

```diff
@@ -0,0 +1,317 @@
+# coding=utf-8
+
+"""
+Collect linux RAID/md state by parsing /proc/mdstat.
+https://raid.wiki.kernel.org/index.php/Mdstat
+
+#### Dependencies
+
+ * /proc/mdstat
+
+"""
+
+import diamond.collector
+import re
+
+
+class MdStatCollector(diamond.collector.Collector):
+    MDSTAT_PATH = '/proc/mdstat'
+
+    def get_default_config_help(self):
+        config_help = super(MdStatCollector, self).get_default_config_help()
+        return config_help
+
+    def get_default_config(self):
+        """
+        Returns the default collector settings
+        """
+        config = super(MdStatCollector, self).get_default_config()
+        config.update({
+            'path': 'mdstat',
+        })
+        return config
+
+    def process_config(self):
+        super(MdStatCollector, self).process_config()
+
+    def collect(self):
+        """Publish all mdstat metrics."""
+        def traverse(d, metric_name=''):
+            for key, value in d.iteritems():
+                if isinstance(value, dict):
+                    if metric_name == '':
+                        metric_name_next = key
+                    else:
+                        metric_name_next = metric_name + '.' + key
+                    traverse(value, metric_name_next)
+                else:
+                    metric_name_finished = metric_name + '.' + key
+                    self.publish_gauge(
+                        name=metric_name_finished,
+                        value=value,
+                        precision=1
+                    )
+
+        md_state = self.parse_mdstat()
+
+        traverse(md_state, '')
+
+    def parse_mdstat(self):
+        """
+        Parse /proc/mdstat.
+
+        File format:
+        The first line is the "Personalities" line.
+        It won't get parsed since it contains only string metrics.
+        The second to second-last lines contain raid array information.
+        The last line contains the unused devices.
+        It won't get parsed since it contains only string metrics.
+
+        :return: Parsed information
+        :rtype: dict
+        """
+
+        try:
+            with open(self.MDSTAT_PATH, 'r') as f:
+                lines = f.readlines()
+        except IOError as err:
+            self.log.exception(
+                'Error opening {mdstat_path} for reading: {err}'.format(
+                    mdstat_path=self.MDSTAT_PATH,
+                    err=err
+                )
+            )
+
+        arrays = {}
+        mdstat_array_blocks = ''
+
+        for line in lines[1:-1]:
```

Comment:
If the IOError is caught, then `lines` is never defined and this line will cause a `NameError`